Fix deprecation warnings related to ORM dialects and add `quarkus.datasource.db-version` as a dialect version hint

yrodiere opened the pull request:
Fixes #31474 

Draft because this PR is based on #31697 , which must be merged first.

In short:

* We stop relying on deprecated, version-specific dialects.
* Instead, we use version-independent dialects and pass a version to Hibernate ORM through configuration properties.
* We set the versionin particular for dialects where the default version in Hibernate ORM is older than the version we were targeting in Quarkus 2.
* In the last commit, we also set the version for other supported dialects. That's not strictly necessary, but it might make sense? Sanne was in favor, personally I'm still debating it.
* We now have tests to check that Hibernate ORM doesn't issue any warning on startup, for all databases. Though only with a rather simple model. This will allow us to detect when ORM no longer supports our default versions, so that's nice.
* We also expose a configuration property `quarkus.datasource.db-version` to set that version explicitly, so that users can now do this instead of setting version-specific dialects.
* When `db-version` is set, we try to check the version on runtime init:
  * if the actual version is higher than or equal to `db-version` we do nothing
  * if the actual version is lower than `db-version` we throw an exception (startup fails)
  * if we can't retrieve the actual version we just log a warning
  * Note this check is only enabled when the version is set explicitly; we don't check default versions because that would break too many applications.

I didn't add documentation beyond the javadoc on the configuration properties (`dialect` and `db-version`) because I'm not sure how useful that would be. I imagine people would mostly end up using this property because they tried a custom dialect and saw a deprecation warning... In which case the javadoc on `dialect` might be enough to point them in the right direction. Thoughts?

yrodiere committed:
Introduce quarkus.datasource.db-min-version and use it in Hibernate ORM

yrodiere committed:
Force the default version of all supported Hibernate ORM dialects

yrodiere committed:
Test with all databases that Hibernate ORM does not log warnings on startup

In particular check there are no dialect-related warnings.

yrodiere committed:
Configure Hibernate ORM with database product versions instead of deprecated, version-specific dialects

yrodiere committed:
Remove the now-unnecessary QuarkusH2Dialect

gsmet commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/config/DialectVersions.java:
```diff
@@ -0,0 +1,47 @@
+ * as H2 is embedded and thus Quarkus determines its default version through the BOM.
+ * See https://github.com/quarkusio/quarkus/issues/1886
+ *
+ * @author Emmanuel Bernard emmanuel@hibernate.org
```
?

yrodiere replied:
Ah right. That's because I originally renamed the `QuarkusH2Dialect` class. Will fix.

gsmet commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/config/DialectVersions.java:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public final class DialectVersions {
```
I think I would have called it `DefaultDialectVersions` to be extra clear but the presence of `toString(DatabaseVersion version)` is not really in line with this.

Maybe rename the constant with `MARIADB_DEFAULT` or something similar?

It might just be me being over cautious but I find the intent not as clear as it should be. Feel free to ignore though if you think otherwise.

yrodiere replied:
I'll wrap this in a static `Defaults` class :)

gsmet commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/config/DialectVersions.java:
```diff
@@ -16,10 +16,13 @@
     public static final String MSSQL = "13"; // 2016
     public static final String MYSQL = "8";
     public static final String ORACLE = "12";
+    public static final String POSTGRESQL = "10.0";
```
I think I would rather defer to ORM than enforcing all versions but that's just my 2 cents.

yrodiere replied:
Yeah, I was on the fence. It seems nobody has a strong opinion though... hard call :)

gsmet reviewed (commented):
Looks very good, thanks. I added a few comments.

yrodiere committed:
Rename quarkus.datasource.db-min-version to db-version

yrodiere committed:
Revert "Force the default version of all supported Hibernate ORM dialects"

We'd rather rely on ORM defaults where possible.

This reverts commit d3b0fd7e0c5dcc860be28e620830a4292550361b.

yrodiere commented:
I rebased, addressed your comments and renamed `db-min-version` to `db-version` as discussed [on Zulip](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/datasource.2Edb-version). Let's see what CI has to say...

gsmet commented:
As a follow-up, I think we should add something in the documentation as it radically changes how users are supposed to handle dialects.
I know the property will be documented in the table but I think a specific paragraph would be welcome to explain how versioned dialects are now working and that you should use `db-version` to point to a specific version.

yrodiere committed:
Document Hibernate ORM dialect configuration

yrodiere commented:
The test failures are unrelated.

I'll go ahead and merge this; feel free to comment on the documentation and I'll send another PR to amend it if necessary.

Thanks for the review!
